quantus/evaluation.py

```diff
@@ -74,94 +79,99 @@ def evaluate(
 
     results: Dict[str, dict] = {}
+    explain_funcs: Dict[str, Callable] = {}
 
-    if isinstance(xai_methods, list):
+    if not isinstance(xai_methods, dict):
+        "xai_methods type is not in: Dict[str, Callable], Dict[str, Dict], Dict[str, np.ndarray]."
 
-        assert a_batch is not None, (
-            "If 'explanation_methods' is a list of methods as strings, "
-            "then a_batch arguments should provide the necessary attributions corresponding "
-            "to each input."
-        )
+    for method, value in xai_methods.items():
 
-        for method in xai_methods:
+        results[method] = {}
 
-            results[method] = {}
+        if callable(value):
 
-            for metric, metric_func in metrics.items():
+            explain_funcs[method] = value
+            explain_func = value
 
-                if progress:
-                    print(f"Evaluating {method} explanations on {metric} metric...")
+            # Asserts.
+            asserts.assert_explain_func(explain_func=explain_func)
 
-                results[method][metric] = agg_func(
-                    metric_func(
-                        model=model,
-                        x_batch=x_batch,
-                        y_batch=y_batch,
-                        a_batch=a_batch,
-                        s_batch=s_batch,
-                        explain_func_kwargs={
-                            **explain_func_kwargs,
-                            **{"method": method},
-                        },
-                        **call_kwargs,
-                    )
-                )
+            # Generate explanations.
+            a_batch = explain_func(
+                model=model,
+                inputs=x_batch,
+                targets=y_batch,
+                **{**explain_func_kwargs, **{"method": method}},
+            )
+            a_batch = utils.expand_attribution_channel(a_batch, x_batch)
 
-    elif isinstance(xai_methods, dict):
+            # Asserts.
+            asserts.assert_attributions(a_batch=a_batch, x_batch=x_batch)
 
-        for method, method_func in xai_methods.items():
+        elif isinstance(value, Dict):
 
-            results[method] = {}
+            if explain_func_kwargs is not None:
+                warnings.warn(
+                    "Passed explain_func_kwargs will be ignored when passing type Dict[str, Dict] as xai_methods."
+                    "Pass explanation arguments as dictionary values."
+                )
 
-            if callable(method_func):
+            explain_func_kwargs = value
+            explain_funcs[method] = explain
 
-                # Asserts.
-                asserts.assert_explain_func(explain_func=method_func)
+            # Generate explanations.
+            a_batch = explain(
+                model=model, inputs=x_batch, targets=y_batch, **explain_func_kwargs
+            )
+            a_batch = utils.expand_attribution_channel(a_batch, x_batch)
```

Comment:
Where do we place abs and normalise operations?

Reply:
You can pre-initialize a metric with `abs` and `normalise` values, and then pass the initialized metric within `metrics` dictionary. a_batch is supposed to get processed in the general_preprocess function accordingly.

Reply:
True! Thanks for that.